packages/transaction-controller/src/utils/history.ts

```diff
@@ -40,16 +56,99 @@ export function updateTransactionHistory(
 
   const currentState = snapshotFromTransactionMeta(transactionMeta);
   const previousState = replayHistory(transactionMeta.history);
-  const historyEntry = generateHistoryEntry(previousState, currentState, note);
+  const newHistoryEntry = generateHistoryEntry(
+    previousState,
+    currentState,
+    note,
+  );
+
+  if (newHistoryEntry.length > 0) {
+    // Casts required here because this list has two separate types of entries:
+    // TransactionMeta and TransactionHistoryEntry. The only TransactionMeta is the first
+    // entry, but TypeScript loses that type information when `slice` is called for some reason.
+    let updatedHistory = transactionMeta.history.slice() as TransactionHistory;
+    updatedHistory.push(newHistoryEntry);
+
+    if (updatedHistory.length > MAX_HISTORY_LENGTH) {
+      updatedHistory = compressTransactionHistory(updatedHistory);
+    }
 
-  if (historyEntry.length > 0) {
     return merge({}, transactionMeta, {
-      history: [...transactionMeta.history, historyEntry],
+      history: updatedHistory,
     });
   }
   return transactionMeta;
 }
 
+/**
+ * Compress the transaction history, if it is possible to do so without compressing entries used
+ * for display. History entries are merged together to make room for a single new entry.
+ *
+ * @param transactionHistory - The transaction history to compress.
+ * @returns A compressed transaction history.
+ */
+function compressTransactionHistory(
+  transactionHistory: TransactionHistory,
+): TransactionHistory {
+  const initialEntry = transactionHistory[0];
+  // Casts required here because this list has two separate types of entries:
+  // TransactionMeta and TransactionHistoryEntry. The only TransactionMeta is the first
+  // entry, but TypeScript loses that type information when `slice` is called for some reason.
+  const historyEntries = transactionHistory.slice(
+    1,
+  ) as TransactionHistoryEntry[];
+
+  const firstNonDisplayedEntryIndex = historyEntries.findIndex(
+    (historyEntry) => {
+      return !historyEntry.some(({ path }) =>
+        ACTIVITY_LOG_HISTORY_PATHS.includes(path),
+      );
+    },
+  );
+
+  // If no non-diplayed entry is found, let history exceed max size.
+  // TODO: Move data used for display to another property, so that we can more reliably limit
+  // history size or remove it altogether.
+  if (firstNonDisplayedEntryIndex === -1) {
+    return transactionHistory;
+  }
+
+  // If a non-displayed entry is found that we can remove, merge it with another entry.
+  // The entry we're merging with might be a "displayed" entry, but that's OK, merging more changes
+  // in does not break our display logic.
```

Comment:
To clarify, are we saying that an entry with the activity paths can have other non-activity fields in, but we don't support multiple activity paths in the same entry?

It would indeed be nice to fix this with a single package upgrade, but is it a trade off since the current UI logic could theoretically be simplified to read the data from the same entry and therefore simplify this compression logic?

Reply:
>but we don't support multiple activity paths in the same entry?

It's not that we don't support multiple paths in the same entry. It's that we only have one timestamp per entry. There's a bit on that here: https://github.com/MetaMask/metamask-extension/blob/c2043772729b52d6f4f4c09f32ead95bdf50b6dc/ui/components/app/transaction-activity-log/transaction-activity-log.util.js#L111

See that we do iterate through every entry, but the comment suggests that each change in an entry "might not" have a timestamp, but the first one always does. If we examine the code here, we can see that the timestamp is present only on the first entry. The timestamp is an important part of how these get displayed.

The goal here was to be non-destructive with displayed data, so I was avoiding merging two displayed entries because that would lose a timestamp.